### PR TITLE
Fix double swap on interrupted revert

### LINKED ISSUE #480: Watchdog reset during bootloader stage breaks image swap
mcuboot version: 1.2.0/v1.3.0-68-g272702a + 0001-MCUBOOT-Config-file-changes.patch
Zephyr version: 1.13.0/zephyr-v1.14.0-834-ga2fa8e9c1c + 0001-ZEPHYR-Add-watchdog-to-smp_svr-sample.patch
Board: nrf52840-Preview-DK
Zephyr SDK: 0.9.5 (for 1.13.0)/0.10.0
mcumgr 0.0.0-dev

I enclose binaries, so that one can build yourself or use mine.
In the patch I've added Watchdog to Zephy's smp_srv sample.

Steps to reproduce:

1. Build boot, application and sign it.
```
$ ./scripts/imgtool.py sign --key root-ec-p256.pem --slot-size 0x69000 --header-size 0x200 --align 8 --version 1.0.0 ../zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/build/zephyr/zephyr.hex zephyr_signed.hex
```
```
$ mergehex -m boot/zephyr/build/zephyr/zephyr.hex zephyr_signed.hex -o zephyr_boot_signed.hex
```

2. Chiperase and flash:

```
$ nrfjprog --program zephyr_boot_signed.hex --chiperase --reset
Parsing hex file.
Erasing user available code and UICR flash areas.
Applying system reset.
Checking that the area to write is not protected.
Programing device.
Applying system reset.
Run.
```

3. Verify
```
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image list
Images:
 slot=0
    version: 1.0.0
    bootable: true
    flags: active confirmed
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
Split status: N/A (0)
```
```
$ nrfjprog --memrd 0x74e58 --n 424
0x00074E58: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074E68: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074E78: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074E88: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074E98: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074EA8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074EB8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074EC8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074ED8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074EE8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074EF8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F08: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F18: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F28: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F38: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F48: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F58: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F68: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F78: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F88: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074F98: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FA8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FB8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FC8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FD8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FE8: FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF   |................|
0x00074FF8: FFFFFFFF FFFFFFFF                     |........|`
```

4. Bump version
```
$ ./scripts/imgtool.py sign --key root-ec-p256.pem --slot-size 0x69000 --header-size 0x200 --align 8 --version 1.0.1 ../zephyrproject/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/build/zephyr/zephyr.bin zephyr_signed.bin
```

5. Update the firmware via mcumgr
```
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image upload /home/mskamra/projects/mcuboot/zephyr_signed.bin
 111.05 KiB / 111.05 KiB [========================================================================================================================================================================================================================================] 100.00% 3.45 KiB/s 32s
Done
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image list
Images:
 slot=0
    version: 1.0.0
    bootable: true
    flags: active confirmed
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
 slot=1
    version: 1.0.1
    bootable: true
    flags: 
    hash: 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
Split status: N/A (0)
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image test 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
Images:
 slot=0
    version: 1.0.0
    bootable: true
    flags: active confirmed
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
 slot=1
    version: 1.0.1
    bootable: true
    flags: pending
    hash: 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
Split status: N/A (0)
```

6. Observe the confirmed flag that stays in slot 1 every reset.
```
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' reset
Done
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image list
Images:
 slot=0
    version: 1.0.1
    bootable: true
    flags: active
    hash: 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
 slot=1
    version: 1.0.0
    bootable: true
    flags: confirmed
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
Split status: N/A (0)
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' reset
Done
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image list
Images:
 slot=0
    version: 1.0.0
    bootable: true
    flags: active
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
 slot=1
    version: 1.0.1
    bootable: true
    flags: confirmed
    hash: 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
Split status: N/A (0)
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' reset
Done
$ mcumgr --conntype ble --connstring 'peer_name=Zephyr' image list
Images:
 slot=0
    version: 1.0.1
    bootable: true
    flags: active
    hash: 99fcb7f25aae5a7657fbf8b2f9a07c464efbc3d6d177ade2ece9b18c3d5ce7ba
 slot=1
    version: 1.0.0
    bootable: true
    flags: confirmed
    hash: 36699217adc9ff5c08a66cd0845a187830977528ccab345480fac6fa54d9c67f
Split status: N/A (0)
```

I enclose the binaries, patches and memory dump 
[wdt_issue.zip](https://github.com/JuulLabs-OSS/mcuboot/files/3187230/wdt_issue.zip)

Here you can observe the peaks after 5 seconds that indicate WDT reset SoC.
VirtualBox_windows10_16_05_2019_15_21_04

### PR DESCRIPTION
**Pull request by ccollins476ad**
This fixes #480.

When mcuboot rewrites image trailers during a swap, some information is lost.  If a reset occurs before the swap completes, mcuboot may not be able to determine which swap type to resume upon startup.  Specifically, if a "revert" swap gets interupted, mcuboot will perform an extraneous swap on the subsequent boot.  See https://github.com/JuulLabs-OSS/mcuboot/issues/480 for details.

**NOTE:** A different PR (#482) adds unit tests which test this fix.

This PR adds an additional field to the image trailer: `swap-type`.  The new trailer structure is illustrated below:

```
     0                   1                   2                   3
     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    ~                                                               ~
    ~    Swap status (BOOT_MAX_IMG_SECTORS * min-write-size * 3)    ~
    ~                                                               ~
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                 Encryption key 0 (16 octets) [*]              |
    |                                                               |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                 Encryption key 1 (16 octets) [*]              |
    |                                                               |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                      Swap size (4 octets)                     |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |   Swap type   |           0xff padding (7 octets)             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |   Copy done   |           0xff padding (7 octets)             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |   Image OK    |           0xff padding (7 octets)             |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
    |                       MAGIC (16 octets)                       |
    |                                                               |
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

[*]: Only present if the encryption option is enabled (`MCUBOOT_ENC_IMAGES`).
```

The `swap-type` field contains one of the `BOOT_SWAP_TYPE_[...]` constants.  Every time a trailer is written, this field is written along with it.  When resuming an interrupted swap, mcuboot uses this field alone to determine the type of swap being resumed. For new swap operations (non-resume case), this field is not read at all; instead, mcuboot consults the `boot_swap_tables` array to determine the swap operation to perform (as it did prior to this PR).

Some additional changes were necessary to make all the simulated unit tests pass:

* Before initiating a new swap operation, always write the image trailer to the scratch area.  This step allows mcuboot to persist the `swap-type` field somewhere before erasing the trailer in the primary slot.  If a reset occurs immediately after the erase, mcuboot recovers by using the trailer in the scratch area.

* Related to the above: if the scratch area is being used to hold status bytes (because there are no spare sectors in the primary slot), erase the scratch area immediately after the trailer gets written to the primary slot.  This eliminates ambiguity regarding the location of the current trailer in case a reset occurs shortly afterwards.

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
I confirm this fix. I've tested this PR with my app from https://github.com/JuulLabs-OSS/mcuboot/issues/480.

**Comment by ccollins476ad** [comment]
Note: I force pushed a commit with documentation and comment changes.  I did not change any code with this force push. (I messed up the trailer structure representation in my first commit).

**Comment by utzig** [comment]
@ccollins476ad Please cherry-pick the commit from PR #482

**Comment by ccollins476ad** [comment]
Thanks for the reminder, @utzig.  I have pulled in your commits from #482.

File: boot/bootutil/src/bootutil_misc.c [hunk 1]
```diff
@@ -119,6 +119,32 @@ boot_flag_decode(uint8_t flag)
...
+ *
+ * @param tbl_val               A magic field from a status source table.
+ * @param val                   The magic value in a trailer, encoded as a
+ *                                  BOOT_MAGIC_[...].
+ *
+ * @return                      1 if the two values are compatible;
+ *                              0 otherwise.
+ */
+int
+boot_magic_eq(uint8_t tbl_val, uint8_t val)
```
Comment (nvlsianpu) [inline boot/bootutil/src/bootutil_misc.c]: Function name is not descriptive. Maybe `boot_magic_compatible _check`

**Comment by utzig** [comment]
@ccollins476ad From your commit message:

> Prior to this change, the scratch image trailer had a different format from a slot image trailer.  Specifically, the `copy_done` field was excluded from the scratch trailer.

More importantly than `copy_done` was that in the scratch area we only ever store a single sector's state, so for it's usage calculation there was: `BOOT_STATUS_STATE_COUNT * min_write_sz`, while for slotX it was: `BOOT_STATUS_MAX_ENTRIES * BOOT_STATUS_STATE_COUNT * min_write_sz`. Now `BOOT_STATUS_MAX_ENTRIES` is user configurable and might end up being a big value; not to say there's anything wrong with the change, but I thought this correction was requried! :-P

**Comment by ccollins476ad** [comment]
@utzig, that is a good point, and something that I completely missed.  At the very least, I will update the commit message.

Do you think this could cause a problem?  I think the scratch area is guaranteed to be large enough to accommodate this change, but I could use a second opinion!

File: boot/bootutil/src/bootutil_misc.c (hunk 1 above)
Comment (ccollins476ad) [reply boot/bootutil/src/bootutil_misc.c]: Thanks, I have renamed it as suggested.  To be honest, I did start out with a longer and more descriptive name, but I had a feeling some of these identifiers were getting too long (the switch from 0/1 to primary/secondary definitely contributes to this!).

**Comment by utzig** [comment]
@ccollins476ad I don't think it will cause issues; so I was thinking about this and it seems to me that if you cannot fit the data of the last sector with metadata on scratch, than it won't fit on the last sector as well, so in the end this seems like a good move into simplifying the code. Also metadata on scratch is only ever used it there's data to be swapped in any slot's last sectors, thus it's a bit of a corner case that would be better off having a specific test added on the simulator. Your changes LGTM, I just need to finish reviewing one commit (the big one!).

File: boot/bootutil/src/bootutil_misc.c [hunk 2]
```diff
@@ -616,6 +642,15 @@ boot_set_pending(int permanent)
             rc = boot_write_image_ok(fap);
         }
 
+        if (rc == 0) {
```
Comment (nvlsianpu) [inline boot/bootutil/src/bootutil_misc.c]: In BOOT_MAGIC_GOOD case swap_type should be checked and checked in case operation was interrupted here.

Comment (utzig) [reply boot/bootutil/src/bootutil_misc.c]: I don't think anyone would be able to schedule a new swap operation if we were resuming an interruption; what exact issue do you have mind?

**Comment by utzig** [comment]
@ccollins476ad Please fixup the last commit.

**Comment by ccollins476ad** [comment]
Thanks, @utzig.  I have fixed up that commit, and reworded the "scratch trailer structure" commit as discussed above.

File: boot/bootutil/src/bootutil_misc.c (hunk 2 above)
Comment (ccollins476ad) [reply boot/bootutil/src/bootutil_misc.c]: I also don't think we need to worry about the resume case here.  `boot_set_pending()` doesn't get called from the boot loader, it gets called from the application.  If the application is running, there can't be a swap in progress.

This is a little weird, though.  If the application tries to schedule a swap operation, and one is already scheduled, we just return 0.  It seems like we ought to check if the two swap operations have the same type.  If they are different, we should return an error.  I think this should go in a separate PR, though.

Comment (nvlsianpu) [reply boot/bootutil/src/bootutil_misc.c]: What I meant was a power-down had occurred just before writing the swap_type. Shouldn't an application be able to recognize such a undone request unambiguously, and be able to renew boot_set_pending() after that? For sure it is possible to check that the application had already collected proper upgrade image.

Comment (ccollins476ad) [reply boot/bootutil/src/bootutil_misc.c]: @nvlsianpu I think I see what you mean.

Application schedules a "test" swap.  This requires two flash writes:
1. Write `magic` to secondary trailer.
2. Write `swap_type` to secondary trailer.

If the device reboots after step 1 (before step 2), the trailer will be incomplete.

However, I don't think this is a problem.  Since this is a new swap (not a resume), mcuboot won't even look at the `swap_type` field at startup.  It should execute the "test" swap as expected.

If mcuboot then reboots in the middle of this swap, then we have a potential issue.  Mcuboot will recognize that a swap was interrupted and needs to be resumed, but it won't be able to determine the swap type.  I think this just puts us back where we were before this bug fix ("revert" will fail; all other swap types happen to work by chance).  Since you cannot specify a "revert" swap type with the `boot_set_pending()` function, I don't think there is a problem here.

I do think there is a general issue of the device resetting with a partially-written trailer.  This issue has always existed, though, so it is probably best left for a separate PR.

Please let me know if I am still missing something.
